File: CPlusPlus/370. 区间加法.cpp
// 基本做法， 将阶梯写成一个数组，排序后挨个处理阶梯间的元素
// 进阶做法： 先将前后之差保存在数组中， 然后做前缀和
#include <algorithm>
#include <vector>

using namespace std;
class BasicSolution {
public:
    vector<int> getModifiedArray(int length, vector<vector<int>>& updates)
    {
        vector<vector<int>> stairs;
        for (const auto& elem : updates) {
            auto start = elem[0];
            auto end = elem[1] + 1;
            stairs.emplace_back(vector<int> { start, elem[2] });
            stairs.emplace_back(vector<int> { end, -elem[2] });
        }
        sort(stairs.begin(), stairs.end());
        int curr_i = 0;
        int curr_val = 0;
        int curr_s = 0;
        vector<int> res(length, 0);
        while (curr_i < res.size() && curr_s < stairs.size()) {
            while (curr_i < res.size() && curr_i < stairs[curr_s][0]) {
                res[curr_i++] = curr_val;
            }
            curr_val += stairs[curr_s++][1];
        }
        return res;
    }
};
class BetterSolution {
public:
    vector<int> getModifiedArray(int length, vector<vector<int>>& updates) {
        vector<int> res(length,0);
        for(const auto& elem:updates){
            res[elem[0]] += elem[2];
            if(elem[1]+1 < length) res[elem[1]+1]-=elem[2];
        }
        for(int i = 1; i < length; ++i){
            res[i] += res[i-1];
        }
        return res;
    }
};

File: Rust/370. 区间加法.rs
// 暴力解法： 居然可以通过
// 优化解法： 先在 res[i] 中记录和前一个位置的差值， 然后做前缀和即为某处的正确结果
impl BruteForceSolution {
    pub fn get_modified_array(length: i32, updates: Vec<Vec<i32>>) -> Vec<i32> {
        let mut res = vec![0; length as usize];
        for i in 0..updates.len() {
            let beg = updates[i][0] as usize;
            let end = updates[i][1] as usize;
            let incr = updates[i][2];
            for j in beg..=end {
                res[j] = res[j] + incr;
            }
        }
        res
    }
}
impl BetterSolution {
    pub fn get_modified_array(length: i32, updates: Vec<Vec<i32>>) -> Vec<i32> {
        let mut res: Vec<i32> = vec![0; length as usize];
        for elem in updates {
            res[elem[0] as usize] = res[elem[0] as usize] + elem[2];
            if elem[1] + 1 < length as i32 {
                res[(elem[1] + 1) as usize] = res[(elem[1] + 1) as usize] - elem[2];
            }
        }
        for i in 1..res.len() {
            res[i] = res[i - 1] + res[i];
        }
        res
    }
}
